src/main/proto/netflix/titus/validator/validator.proto

```diff
@@ -0,0 +1,200 @@
+    uint32 loadTime = 2;
+
+    // Data expiry time.
+    uint32 expiryTime = 3;
+
+    oneof Decision {
+        Authoritative authoritative = 4;
+
+        Authoritative nonAuthoritative = 5;
+    }
+}
+
+message ValidationFailure {
+
+    // Validation rule specific error code. Common error codes:
+    // * 'notFound' - resource not found
+    // * 'accessDenied' - access denied to the resource
+    // * 'resourceNotAvailable' - connectivity to external resource failure
+    string errorCode = 1;
+
+    // Validation failure details.
+    string errorMessage = 2;
+}
+
+message SecurityGroupValidationRequest {
+    // A security group value can be:
+    // * a security group id (for example sg-12345678)
+    // * a security group ARN
+    // * a security group name/alias (for example 'titus-sg')
+    string securityGroup = 1;
+
+    // A subnet id associated with the given security group. If not set, a default subnet for the given account is used.
+    string subnetId = 2;
+}
+
+message SecurityGroupValidationResponse {
+
+    message Success {
+        // An alias associated with the given security group or empty string if not defined.
+        string securityGroupAlias = 1;
+
+        // Security group id.
+        string securityGroupId = 2;
+
+        // Security group ARN.
+        string securityGroupArn = 3;
+    }
+
+    oneof Result {
+        Success success = 1;
+
+        ValidationFailure failure = 2;
+    }
+
+    CacheMetadata cacheMetadata = 3;
+}
+
+message IamRoleValidationRequest {
+    // IAM role name or ARN.
+    string iamRole = 1;
+}
+
+message IamRoleValidationResponse {
+
+    message Success {
+        // IAM role name.
+        string iamRoleName = 1;
+
+        // ARN associated with the give IAM role.
+        string iamRoleArn = 2;
+    }
+
+    oneof Result {
+        Success success = 1;
+
+        ValidationFailure failure = 2;
+    }
+
+    CacheMetadata cacheMetadata = 3;
+}
+
+message S3BucketAccessValidationRequest {
+
+    // A bucket identifier, which can be:
+    // * a bucket name
+    // * a bucket ARN
+    // * a bucket access point
+    string bucket = 1;
+
+    // A bucket prefix
+    string bucketPrefix = 2;
+
+    // IAM role for which access is checked.
+    string iamRole = 3;
+}
+
+message S3BucketAccessValidationResponse {
+
+    message Success {
+        // True if a bucket can be read.
+        bool readAccess = 1;
+
+        // True if a bucket can be written to.
+        bool writeAccess = 2;
+    }
+
+    oneof Result {
+        Success success = 1;
+
+        ValidationFailure failure = 2;
+    }
+
+    CacheMetadata cacheMetadata = 3;
+}
+
+// One of name/tag or digest is expected. If both are set, the name/tag is validated first, and its digest is compared
+// against the provided one. All checks must succeed for the validation to pass.
+message ImageValidationRequest {
+
+    // Image name.
+    string name = 1;
+
+    // Image tag (required if name is set).
+    string tag = 2;
```

Comment:
Why is the tag required if name is set? A digest should be a valid substitute.

Reply:
Changing to:
```
// Image name is always required. One of tag or digest must be always set. If both tag and digest is set, the
// digest resolved for the tag must match the provided digest value.
```